docs/spec_draft.md

```diff
@@ -1810,6 +1811,86 @@ where `i[d] = j[permutation[d]]`.
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.triangular_solve
+
+### Semantics
+
+Solves systems of linear equations of a batch of matrices with lower or upper
+triangular coefficients by forward- or back-substitution.
+
+More formally, given `a` and `b`, `result[i0, ..., iR-3, :, :]` is the solution
+to `op(a[i0, ..., iR-3, :, :]) * x = b[i0, ..., iR-3, :, :]` when `left_side` is
+`true` or `x * op(a[i0, ..., iR-3, :, :]) = b[i0, ..., iR-3, :, :]` when
+`left_side` is `false`, solving for the variable `x` where `op(a)` is either
+`op(a) = a`, or `op(a) = Transpose(a)`, or `op(a) = Conj(Transpose(a))`.
+
+Input data is read only from the lower triangle of `a`, if `lower` is `true` or
+upper triangle of `a`, otherwise. Output data is returned in the same triangle;
+the values in the other triangle are implementation-defined.
+
+If `unit_diagonal` is true, the diagonal elements of `a` are assumed to be 1 and
+not accessed.
+
+`transpose_a` can be one of the following:
+  * `TRANSPOSE_INVALID`: Illegal matrix; transpose is invalid.
+  * `NO_TRANSPOSE`: Perform operation using `a` as-is.
+  * `TRANSPOSE`: Perform operation on transpose of `a`.
+  * `ADJOINT`: Perform operation on conjugate transpose of `a`.
+
+### Inputs
+
+| Name            | Type                                                                    |
+|-----------------|-------------------------------------------------------------------------|
+| `a`             | tensor of floating-point or complex type                                |
+| `b`             | tensor of floating-point or complex type                                |
+| `left_side`     | constant of type `i1`                                                   |
+| `lower`         | constant of type `i1`                                                   |
+| `unit_diagonal` | constant of type `i1`                                                   |
+| `transpose_a`   | enum of `TRANSPOSE_INVALID`, `NO_TRANSPOSE`, `TRANSPOSE`, and `ADJOINT` |
+
+### Outputs
+
+| Name     | Type                                     |
+|----------|------------------------------------------|
+| `result` | tensor of floating-point or complex type |
+
+### Constraints
+
+  * (C1) `a`, `b`, and `result` have the same element type and rank $\ge$ 2.
+  * (C2) dim(`a`, -2) = dim(`a`, -1).
+  * (C3) dim(`a`, `i`) $=$ dim(`b`, `i`) for all `i` $\in$ [0, R-3]. If
```

Comment:
I preferred the following better but might be less readable.
 dim(`a`, i) $=$ dim(`b`, i) if i $\in$ [0, R-3] and dim(`a`, i) $=$ dim(`b`, R - `left_side` ? 2 : 1 ) otheriwse.

Let us keep your edit.

Reply:
How about the following reformulation:

```
  * (C3) The relationship between `shape(a)` and `shape(b)` is as follows:
    * For all `i` $\in$ [0, R-3], dim(`a`, `i`) $=$ dim(`b`, `i`).
    * `dim(a, R-2)` = `dim(b, left_side ? R-2 : R-1)`.
    * `dim(a, R-1)` is the same as discussed in C2.
```

The idea is to emphasize that C3 exhaustively covers all dimension sizes of `a`. The current formulation makes it hard because all the cases are mixed together in the same paragraph together with an if. Also, `dim(a, ...)` is on the left-hand side of equations in the first sentence, and on the right-hand side in the second sentence.

Reply:
`dim(a, R-1)` = `dim(a, R-2)` = `dim(b, left_side ? R-2 : R-1)`.

How about we remove C2 and the third sub point and combine them like this?

Reply:
Keeping the rest of C3 and the first sub-point as is ofcourse

Reply:
Sure thing.